File: scala-procedural/cipher.scala
// Caesar Cipher - Scala (procedural)
// Author: John Paul Welsh
object Cipher {
    def encrypt(inputStr: String, shiftAmt: Int):String = {
        var outputStr = ""
        var newCh = ' '
        var newCode = 0

        for(chCount <- 0 until 15) {
            newCh = inputStr.charAt(chCount)
            newCode = newCh.asInstanceOf[Int]
            if (newCode != 32) {
                newCode = newCode + shiftAmt
                if (newCode > 90) newCode = 64 + (newCode % 90)
                if (newCode < 65) newCode = 91 - (65 % newCode)
                newCh = newCode.asInstanceOf[Char]
            }
            outputStr = outputStr + newCh
        }

        return outputStr
    }

    def decrypt(inputStr: String, shiftAmt: Int):String = {
        var outputStr = ""
        var newCh = ' '
        var newCode = 0

        for(chCount <- 0 until 15) {
            newCh = inputStr.charAt(chCount)
            newCode = newCh.asInstanceOf[Int]
            if (newCode != 32) {
                newCode = newCode - shiftAmt
                if (newCode > 90) newCode = 64 + (newCode % 90)
                if (newCode < 65) newCode = 91 - (65 % newCode)
                newCh = newCode.asInstanceOf[Char]
            }
            outputStr = outputStr + newCh
        }

        return outputStr
    }

    def solve(inputStr: String, maxShiftVal: Int) {
        var shift = 0
        var newCh = ' '
        var newCode = 0

        for(iteration <- maxShiftVal to 0 by -1) {
            var outputStr = ""
            shift = iteration
            
            for(chCount <- 0 until 15) {
                newCh = inputStr.charAt(chCount)
                newCode = newCh.asInstanceOf[Int]
                if (newCode != 32) {
                    newCode = newCode - shift
                    if (newCode > 90) newCode = 64 + (newCode % 90)
                    if (newCode < 65) newCode = 91 - (65 % newCode)
                    newCh = newCode.asInstanceOf[Char]
                }
                outputStr = outputStr + newCh
            }
            println("Caesar " + iteration + ": " + outputStr)
        }
    }

    def main(args: Array[String]) {
        var inputStr = "John Paul Welsh"
        var shiftAmt = 1
        var maxShiftVal = 26
        
        inputStr = inputStr.toUpperCase()

        println("The input string is " + inputStr)
        println("The shift amount for encrypt and decrypt is " + shiftAmt)
        println("The maximum shift value for solve is " + maxShiftVal)
        println("Encrypted: " + encrypt(inputStr, shiftAmt))
        println("Decrypted: " + decrypt(inputStr, shiftAmt))
        solve(inputStr, maxShiftVal)
    }
}

File: pascal/cipher.pas
{Caesar Cipher - Pascal}
{Author: John Paul Welsh}
Program cipher;

Var input : String;
Var shiftAmount, maxShiftVal: Integer;

{-------------------------------------------------------}
Procedure encrypt(input : String; shiftAmount : Integer);

Var output : String;
Var strPos, ascii, charCount : Integer;

Begin
    output := '';

    For strPos := 1 to 15 do
    Begin
        ascii := ord(input[strPos]);
        
        if not(ascii = 32) then
        Begin
            ascii := ascii + shiftAmount;
            if (ascii > 90) then ascii := 64 + (ascii mod 90);
            if (ascii < 65) then ascii := 91 - (65 mod ascii);
        End;
        
        output := output + chr(ascii);
    End;
    
    Writeln('Encrypt: ', output);
End;
{-------------------------------------------------------}

{-------------------------------------------------------}
Procedure decrypt(input : String; shiftAmount : Integer);

Var output : String;
Var strPos, ascii, charCount : Integer;

Begin
    output := '';

    For strPos := 1 to 15 do
    Begin
        ascii := ord(input[strPos]);
        
        if not(ascii = 32) then
        Begin
            ascii := ascii - shiftAmount;
            if (ascii > 90) then ascii := 64 + (ascii mod 90);
            if (ascii < 65) then ascii := 91 - (65 mod ascii);
        End;
        
        output := output + chr(ascii);
    End;
    
    Writeln('Decrypt: ', output);
End;
{-------------------------------------------------------}

{-------------------------------------------------------}
Procedure solve(input : String; maxShiftVal : Integer);

Var output : String;
Var strPos, ascii, charCount, shiftAmount, iteration : Integer;

Begin
    For iteration := maxShiftVal DOWNTO 0 do
    Begin
        output := '';
        shiftAmount := iteration;
        
        For strPos := 1 to 15 do
        Begin
            ascii := ord(input[strPos]);
            
            if not(ascii = 32) then
            Begin
                ascii := ascii + shiftAmount;
                if (ascii > 90) then ascii := 64 + (ascii mod 90);
                if (ascii < 65) then ascii := 91 - (65 mod ascii);
            End;
            
            output := output + chr(ascii);
        End;
        
        Writeln('Caesar ', iteration, ': ', output);
    End;
End;
{-------------------------------------------------------}

{-------------------------------------------------------}
Begin
    input := 'John Paul Welsh';
    shiftAmount := 12;
    maxShiftVal := 30;
    
    Writeln('The input string is ', input);
    Writeln('The shift amount for encrypt and decrypt is ', shiftAmount);
    Writeln('The maximum shift value for solve is ', maxShiftVal);
    
    input := upcase(input);
    
    shiftAmount := shiftAmount mod 26;
    
    encrypt(input, shiftAmount);
    decrypt(input, shiftAmount);
    solve(input, maxShiftVal);
End.
{-------------------------------------------------------}
